src/index.ml

```diff
@@ -82,7 +82,12 @@ module type S = sig
 
   val iter : (key -> value -> unit) -> t -> unit
 
-  val force_merge : ?hook:[ `After | `Before ] Private.Hook.t -> t -> unit
+  type async_type
+
+  val force_merge :
+    ?hook:[ `After | `Before ] Private.Hook.t -> t -> async_type option
+
+  val join : async_type option -> unit
```

Comment:
I'm a bit reluctant to having this in the public API, maybe it could be moved to `Private`?

Reply:
Do you not like the notion of being able to bind on the `force_merge` finishing at all, or just think it's a bit ugly? I think it might be prettier as:
```ocaml
type async

val force_merge : ?hook:[ `After | `Before ] Private.Hook.t -> t -> unit async
val await : 'a async -> 'a
```

Reply:
(or switch `async`/`await` to `promise`/`resolve` if you're the web-developing type)

Reply:
@pascutto I leave the `Private` part to https://github.com/mirage/index/pull/147 then 

@CraigFe you are right, I can hide the option in the type `async` as well (done now), but I'm not sure why we should have a parametric type here.

Reply:
No good reason :slightly_smiling_face: I suggested it because to me the type `'a t -> 'a` strongly suggests the implementation "_block on the resolution of 'a-valued promise_", by analogy to:

|    |   |   |
| -- | -- | -- |
| OCaml | [`Lwt_main.run`](https://ocsigen.org/lwt/4.4.0/api/Lwt_main) | `'a t -> 'a` |
| Haskell | [`wait`](http://hackage.haskell.org/package/async-2.2.2/docs/Control-Concurrent-Async.html#v:wait) | `Async a -> IO a` |
| Rust  | [`join`](https://doc.rust-lang.org/std/thread/struct.JoinHandle.html#methods) | `JoinHandle<A> -> Result<A>` |

whereas `t -> unit` doesn't speak to me as much (e.g. it might do cancellation like `Lwt.cancel : 'a t -> unit` or `cancel :: Async a -> IO ()`). But on closer inspection our Unix threads are old-fashioned value-less ones anyway :stuck_out_tongue: Sorry!

Reply:
(it also makes effectively no difference if the type is private anyway...)

Reply:
I agree that the type `'a t ->'a` is more suggestive. I propose we leave like this for now, and the PR that uses Lwt will change a lot of things anyway and we can change this too then.